### PR TITLE
Add a const-constructors for sync primitives

### LINKED ISSUE #2756: Const constructors for Semaphore, Mutex, etc. with parking_lot.
I’m noticing that with the `parking_lot` feature enabled, the type `tokio::sync::Semaphore` is made up of types that can all be constructed with `const fn`s. Putting a `Mutex` into a `static` variable is a common pattern to simulate global variables in Rust; without something like `const fn new` available, one also needs to wrap the variable into a `lazy_static` or something similar. Thus the following seems like a valuable addition to me.

I propose that a few functions are made conditionally `const`, when the `parking_lot` feature is enabled. I’m not sure yet how nice this can work out in the documentation and how bad it would look in the source code though. If making e.g. `Mutex::new` conditionally `const` causes too much churn, an alternative is to add some _additional_ functions, gated behind the `parking_lot` feature, for `const` construction of `Semaphore`, `Mutex`, etc.

I have no idea what this `loom` feature is all about and how it interacts with this feature request.
<hr>

As far as I can tell, we would just need to add const to
* `loom::std::parking_lot::Mutex::new`
* `util::LinkedList::new`
* `sync::batch_semaphore::Semaphore::new` (conditionally)
* `sync::Semaphore::new` (conditionally)
* `sync::Mutex::new` (conditionally)
* `sync::RwLock::new` (conditionally)
* _possibly other users of `Semaphore` I haven’t found yet._ (Edit: for example `Notify`.)

### PR DESCRIPTION
**Pull request by mental32**
## Motivation

#2756 describes that the use of const-constructors with sync primitives and that there are currently no const-constructors for them.

Closes #2756 

## Solution

I've added a `const_new` public function to:

  * `loom::std::atomic_usize::AtomicUsize`
  * `loom::std::parking_lot::Mutex`
  * `sync::batch_semaphore::Semaphore`
  * `sync::mutex::Mutex`
  * `util::linked_list::LinkedList`

All of the `const_new` constructors have the following cfg:

```rust
    #[cfg(all(
        feature = "parking_lot", 
        not(all(loom, test)),
    ))]
```

As specified by [the concern in this comment](https://github.com/tokio-rs/tokio/issues/2756#issuecomment-672119453)

## Disclaimer

I have not modified or added any new tests as I'm not sure what needs to be done on the testing side for these changes. also since the cfg specifies that the constructors should not be present when `cfg(test)`

### CODE REVIEW

File: tokio/src/sync/mutex.rs
Comment (MikailBag) [inline tokio/src/sync/mutex.rs:229]:
```suggestion
    /// static lock: Mutex<i32> = Mutex::const_new(5);
```

**Comment by mental32** [comment]
There are some CI tests that I don't believe will work, due to a using `feature(const_fn)` this feature is however needed since if I try to build it on stable without it I get the following error:

```
   Compiling tokio v0.3.0 (/home/mental/Desktop/open_projects/tokio/tokio)
error[E0723]: trait bounds other than `Sized` on const fn parameters are unstable
  --> tokio/src/util/linked_list.rs:69:6
   |
69 | impl<T: Link> LinkedList<T> {
   |      ^
   |
   = note: see issue #57563 <https://github.com/rust-lang/rust/issues/57563> for more information
   = help: add `#![feature(const_fn)]` to the crate attributes to enable

error: aborting due to previous error

For more information about this error, try `rustc --explain E0723`.
error: could not compile `tokio`.

To learn more, run the command again with --verbose.
```

I'm not sure how to specify the cfg this PR so that these changes are only available on nightly.

**Comment by mental32** [comment]
> I'm not sure how to specify the cfg this PR so that these changes are only available on nightly.

never mind, figured it out :sweat_smile:

File: tokio/src/sync/mutex.rs
Comment (Darksonn) [inline tokio/src/sync/mutex.rs:231]: Why does it require nightly?

Comment (mental32) [reply tokio/src/sync/mutex.rs:231]: `Mutex` uses `Semaphore` uses `LinkedList`, `LinkedList` has a trait bound which when using the const-constructor requires `feature(const_fn)` which is nightly see [this comment](https://github.com/tokio-rs/tokio/pull/2790#issuecomment-679164335)

Comment (mental32) [reply tokio/src/sync/mutex.rs:231]: I can only see https://github.com/rust-lang/rfcs/pull/2632 as the closest changeset that might let us relieve using `feature(const_fn)`

File: tokio/src/util/linked_list.rs
Comment (hawkw) [inline tokio/src/util/linked_list.rs:79]: We wouldn't need the nightly feature gate if we changed the `LinkedList` type definition so that we don't need the `T: Link` bound.

Currently, it's necessary because a linked list is defined as https://github.com/tokio-rs/tokio/blob/3d53e15328c77031af8c30c6abd1eae7e772e64a/tokio/src/util/linked_list.rs#L15-L21
where the `T::Target` type is used. However, we _could_ change the definition to
```rust
pub(crate) struct LinkedList<L, T> {
    /// Linked list head
    head: Option<NonNull<T>>,

    /// Linked list tail
    tail: Option<NonNull<T>>,

    /// Node type marker
    _link: PhantomData<fn(L)>,
}
```

put the `const_new` function in an unconstrained `impl` block, like
```rust
impl<L, T> LinkedList<L, T> {
    pub(crate) const fn const_new() -> LinkedList<T> {
        LinkedList {
            head: None,
            tail: None,
        }
    }
}
```
and all the other functions (which _do_ require the `Link` bound) in a block with
```rust
impl<L: Link> LinkedList<L, L::Target> {
    // ...
}
```

This would be a much bigger change, since it would also require changing all the uses of `LinkedList`...but, it would let us have the const fn constructors on stable.

@carllerche, do you agree that not needing nightly Rust for const fn constructors is worth a somewhat larger refactoring of the linked list implementation?

File: tokio/src/loom/std/parking_lot.rs
Comment (hawkw) [inline tokio/src/loom/std/parking_lot.rs:30]: let's add
```suggestion
    #[cfg(all(feature = "nightly", feature = "parking_lot", not(all(loom, test)),))]
    #[cfg_attr(docsrs, doc(cfg(all(feature = "parking_lot", feature = "nightly"))))]
```
so that the  required features show up in the RustDoc output?

File: tokio/src/sync/mutex.rs
Comment (hawkw) [inline tokio/src/sync/mutex.rs:231]: let's add the `doc_cfg` attr here as well:
```suggestion
    #[cfg(all(feature = "nightly", feature = "parking_lot", not(all(loom, test)),))]
    #[cfg_attr(docsrs, doc(cfg(all(feature = "parking_lot", feature = "nightly"))))]
```

File: tokio/src/util/linked_list.rs
Comment (Darksonn) [reply tokio/src/util/linked_list.rs:79]: I would certainly prefer this to adding a nightly feature flag.

Comment (mental32) [reply tokio/src/util/linked_list.rs:79]: I'm writing this without looking too deeply at the internals but if a refactor wouldn't be worth it couldn't we have a duplicate `LinkedList` defined like above and then use that for semaphore's internal linked list?

Comment (hawkw) [reply tokio/src/util/linked_list.rs:79]: > I'm writing this without looking too deeply at the internals but if a refactor wouldn't be worth it couldn't we have a duplicate `LinkedList` defined like above and then use that for semaphore's internal linked list?

I think the refactor is pretty mechanical, and IMO, maintaining two duplicate implementations of unsafe code like the linked list feels higher-risk than just making the change everywhere.

**Comment by mental32** [comment]
I'm not married to using a nightly feature flag or nightly at all for that matter so currently I'm open to any kind change :)

As for moving forward with this, I'm not quite sure what the next steps are.. Currently I'm watching this thread waiting for other maintainers opinions before doing anything so I feel like this is just blocked still atm.

**Comment by Darksonn** [comment]
Can you open a new PR that makes the change we described to `LinkedList`? Then we'll merge that and continue here.

**Comment by hawkw** [comment]
@mental32 btw, feel free to ping me if you get stuck on the `LinkedList` change. Thanks!

**Comment by mental32** [comment]
Sure thing! It may take a little while however so bear with me :)

**Comment by mental32** [comment]
All ready :+1:  only blocked on #2805 for the moment

**Comment by Darksonn** [comment]
You should be able to merge in master now with the new `LinkedList`. Please prefer merges to rebases.

**Comment by mental32** [comment]
I think everything's good to go here :smile:

File: tokio/src/loom/std/atomic_usize.rs
Comment (Darksonn) [inline tokio/src/loom/std/atomic_usize.rs:23]: Can we not just add the `const` marker to the ordinary `new` function in this case? It is an internal type, so it introduces no backwards compatibility challenges, and the functions are identical.

File: tokio/src/sync/batch_semaphore.rs
Comment (Darksonn) [inline tokio/src/sync/batch_semaphore.rs:131]: Can the missing assertion cause trouble? E.g. do we have unsafe code that relies on it?

Comment (mental32) [reply tokio/src/sync/batch_semaphore.rs:131]: Any code that would've been relying on the assertion would have already been using the alternate `new` constructor, so maybe a note in the doc-string about safety is warranted here?

Comment (mental32) [reply tokio/src/sync/batch_semaphore.rs:131]: tldr; no I don't think there's anything that would cause trouble here

Comment (Darksonn) [reply tokio/src/sync/batch_semaphore.rs:131]: > Any code that would've been relying on the assertion would have already been using the alternate new constructor.

That's not really what I mean. We now have a safe function that allows creation of a semaphore with too many permits. Does that make the API unsound, i.e. does it expose a way of using it that causes UB using only the safe interface?

For example, let's say someone comes along and adds a `const fn new` to `Semaphore`. Could that cause undefined behavior?

Comment (mental32) [reply tokio/src/sync/batch_semaphore.rs:131]: I don't know enough about how the permits function or what they're used for to answer that unfortunately, perhaps someone else can chime in?

Comment (carllerche) [reply tokio/src/sync/batch_semaphore.rs:131]: We do need to ensure `permits` does not exceed the max number of permits for it to be safe. We can do this by saturating probably. If `permits` is greater than `MAX_PERMITS`, then limit it to `MAX_PERMITS`. This can be documented by the fn.

Comment (mental32) [reply tokio/src/sync/batch_semaphore.rs]: I made a change where a guard was introduced to clamp `permits` to `MAX_PERMITS` when exceeded. I did not realize however that if's aren't allowed in const fn's on stable yet so this change is now moot.

Should I mark the function `unsafe` and document the requirements under a `safety` section in the doc-comment instead? @carllerche

Comment (mental32) [reply tokio/src/sync/batch_semaphore.rs]: I've investigated it a little more after observing the other tests passing and it looks like [this test](https://github.com/tokio-rs/tokio/pull/2790/checks?check_run_id=1089992522) is an outlier with an out-of-date stable version `1.45.2` where the const-fn's can't have if's in them but as of stable `1.46.0` they can.

This raises the question, should the test be fixed to use the newer stable version and this usage of this feature now renders a minimal supported rust version of `1.46.0` or should an alternative be applied like described above with marking the constructor `unsafe` and not require an update ?

Comment (Darksonn) [reply tokio/src/sync/batch_semaphore.rs]: You could do the following since `MAX_PERMITS` is of the form `2^n - 1`.
```
permits = permits & Self::MAX_PERMITS;
```

**Review by Darksonn — APPROVED** [review]
I will merge it once CI passes.

**Pull request merged**
